### PR TITLE
Update: no-sequences requires parens in assignment position

### PR DESCRIPTION
**Pull request by kaicataldo**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [x] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Tell us about your environment**

* **ESLint Version:** `master`
* **Node Version:** N/A
* **npm Version:** N/A

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**
<details>
<summary>Configuration</summary>

```json
{
    "rules": {
        "no-sequences": "error"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
var a = (b = 1, c = 2);
a = (b = 1, c); 
```

**What did you expect to happen?**

I expected the following to error due to the assignment being a sequence expression.

```js
var a = (b = 1, c = 2);
a = (b = 1, c); 
```

The above examples should warn unless they are wrapped in double parens, due to the single parens changing the semantics of this statement.

**What actually happened? Please include the actual, raw output from ESLint.**

[Demo example](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2VxdWVuY2VzOiBcImVycm9yXCIqL1xuXG4vLyBUaGlzIGlzIGEgc2VxdWVuY2UgYW5kIHNob3VsZCB3YXJuXG52YXIgYSA9IChiID0gMSwgYyA9IDIpIC8vIGEgPT09IDI7XG5cbi8vIFRoaXMgc2hvdWxkIG5vdCB3YXJuXG52YXIgYSA9IGIgPSAxLCBjID0gMiAvLyBhID09PSAxIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

```js
// Should error
var a = (b = 1, c = 2);  // No error
a = (b = 1, c);  // No error
a = b = 1, c; // Error

// Should not error
var a = ((b = 1, c = 2));  // No error
a = ((b = 1, c));  // No error
var a = b = 1, c = 2;  // No error
var a = b = 1, c;  // No error
```

#### What changes did you make? (Give an overview)

I updated the rule to check sequence expressions that are assigned as values.

#### Is there anything you'd like reviewers to focus on?

Would love to verify that my analysis of what this rule should and shouldn't warn on seems correct to everyone else. This case in particular:

```js
// Should not warn because it is a declaration and can not be used an expression
var a = b = 1, c = 2; 
var a = b = 1, c; 

// Should warn because it is just an assignment and can be used as an expression
a = b = 1, c = 2;
a = b = 1, c;
```

### CODE REVIEW

File: lib/rules/no-sequences.js [hunk 1]
```diff
@@ -55,15 +55,39 @@ module.exports = {
...
         };
 
+        /**
+         * Determines whether a node is a SequenceExpression being assigned as
+         *     a value while containing an AssignmentExpression.
+         * @param {ASTNode} node The AST node.
+         * @returns {boolean} True if sequence containing assignments is being
+         *     assigned as a value.
+         */
+        function isAssignmentSequenceInAssignment(node) {
```
Comment (kaicataldo) [inline lib/rules/no-sequences.js]: Not convinced about this function name - definitely open to suggestions :)

File: tests/lib/rules/no-sequences.js [hunk 2]
```diff
@@ -65,6 +85,25 @@ ruleTester.run("no-sequences", rule, {
...
+        { code: "let a = (b, c = 2)", env: { es6: true }, errors: errors(11) },
+        { code: "let a = (b = 1, c)", env: { es6: true }, errors: errors(15) },
+        { code: "const a = (b = 1, c = 2)", env: { es6: true }, errors: errors(17) },
+        { code: "const a = (b, c = 2)", env: { es6: true }, errors: errors(13) },
+        { code: "const a = (b = 1, c)", env: { es6: true }, errors: errors(17) },
+        { code: "const a = (b = c = 1, d)", env: { es6: true }, errors: errors(21) },
+        { code: "a = (b = 1, c = 2)", errors: errors(11) },
+        { code: "a = (b, c = 2)", errors: errors(7) },
+        { code: "a = (b = 1, c)", errors: errors(11) },
+        { code: "a = b = 1, c = 2", errors: errors(10) },
```
Comment (kaicataldo) [inline tests/lib/rules/no-sequences.js]: Actually, I think we can make this check smarter by checking if it's a child of an `ExpressionStatement` or not. Will update.

**Comment by mdjermanovic** [comment]
It seems that the purpose of this rule is to warn on accidental use of the comma operator:

> However, this frequently obscures side effects, and its use is often an accident. 

Off-topic, then maybe the title in the docs `Disallow Use of the Comma Operator` isn't quite correct. In [When Not To Use It](https://eslint.org/docs/rules/no-sequences#when-not-to-use-it) is described how to actually disallow comma operator.

```js
// Should error
var a = (b = 1, c = 2);  // No error
a = (b = 1, c);  // No error
```

This is a tough question, parens are required here for the semantics, but they also represent intention.

**Comment by mdjermanovic** [comment]
Why is just the case where `=` is both inside and outside a special case, parens often change semantics with `,`:

```js
x = (f(), g());
```

**Comment by mdjermanovic** [comment]
If the purpose of this rule is to warn on possible accidental use of `,` operators:

> ```js
> a = (b = 1, c);  // No error
> ```

this doesn't look like it was supposed to be something else, especially because of the `=` inside. It could be a pattern to make sure that `a` and `b` are set at the same time?

I'd say that the following looks more like an error:

> ```js
> a = (b, c);  // No error
> ```

These parens change semantics in the same way as in the first example, but the use of `,` seems to be meaningless here because `b` has no side effects. This code looks like someone forgot to call a function.

Maybe there should be an option to warn if any of the `,` operands doesn't have side effects (only the last one would be allowed not to have).

**Comment by kaicataldo** [comment]
I see your point. I was interpreting this part of the [docs](https://eslint.org/docs/rules/no-sequences#rule-details) as requiring extra parens to explicitly mark the usage of comma operators (and therefore requiring double parens), but I can see how it could also be valid to say that, because it already has parens, that is already explicit usage of the comma operator.

**Comment by kaicataldo** [comment]
Thanks for pointing this out - having spent more time thinking about this, I think your interpretation is correct. I should have followed my own advice and made an issue for this before implementing it 😅 

Thanks for the discussion!

**Comment by mdjermanovic** [comment]
I just didn't see `=` as a special case in that context, but I do think that requiring double parens might be a good idea to evaluate (maybe to require always, if an option is enabled).

It looks like `no-sequences` is for people who actually want to use sequences sometimes :) (otherwise, there's `no-restricted-syntax`). Maybe someone wants to use `,` only in for-loops, so warning on a single pair of parens might be useful to avoid mistakes.

**Comment by kaicataldo** [comment]
> I just didn't see = as a special case in that context ...

Oh yeah, no worries! Given the discussion above about how the commas in multi variable declarations, I thought it was a special case because the parens aren't optional in that case (i.e. changing from a multi variable declarations to a single variable declaration with a sequence as the value).

> It looks like no-sequences is for people who actually want to use sequences sometimes :) 

Yes, it does seem like the rule's name and description aren't completely accurate. Seems like it should be named something like `no-ambiguous-sequence` or `no-confusing-sequence` (similar to [no-confusing-arrow](https://eslint.org/docs/rules/no-confusing-arrow)).

**Pull request closed without merging**